// epitech-api/build/connector.js
(function() {
  var EpitechAPIConnector;

  EpitechAPIConnector = (function() {
    function EpitechAPIConnector(login, password, stream) {
      var EpitechAPISettings, request;
      this.login = login;
      this.password = password;
      this.isSignedIn = false;
      EpitechAPISettings = require("./settings").EpitechAPISettings;
      this.settings = new EpitechAPISettings;
      this.stream = stream;
      request = require("request");
      this._cookieJar = request.jar();
      this.request = require("request").defaults({
        jar: this.cookieJar
      });
    }

    EpitechAPIConnector.prototype.log = function(string) {
      if (this.stream) {
        return this.stream.write("" + string + "\n");
      }
    };

    EpitechAPIConnector.prototype.formatRequest = function(response) {
      return JSON.parse.toString(response.body.substr(response.body.indexOf('\n')));
    };

    EpitechAPIConnector.prototype.signIn = function(callbackSuccess, callbackFailure) {
      var data, url;
      data = {
        form: {
          login: this.login,
          password: this.password,
          remind: true
        }
      };
      url = this.settings.url_signIn();
      this.log("Getting " + url);
      return this.request.post(url, data, (function(_this) {
        return function(error, response) {
          var json;
          if (!error && response.statusCode === 200) {
            _this.isSignedIn = true;
            json = _this.formatRequest(response);
            _this.log("Student has been logged in");
            if (callbackSuccess) {
              callbackSuccess(json);
            }
          } else {
            _this.log("Student cannot be logged in");
            if (callbackFailure) {
              callbackFailure(error, response);
            }
          }
          return _this.isSignedIn;
        };
      })(this));
    };

    EpitechAPIConnector.prototype.getModule = function(year, codemodule, codeinstance, callbackSuccess, callbackFailure) {
      var url;
      url = this.settings.url_module(year, codemodule, codeinstance);
      this.log("Getting " + url);
      return this.request.get(url, (function(_this) {
        return function(error, response) {
          var json;
          if (!error && response.statusCode === 200) {
            json = _this.formatRequest(response);
            _this.log("Succesfully get'd " + url);
            if (callbackSuccess) {
              return callbackSuccess(json);
            }
          } else {
            _this.log("Failed to get " + url);
            if (callbackFailure) {
              return callbackFailure(error, response);
            }
          }
        };
      })(this));
    };

    EpitechAPIConnector.prototype.getMeetingSlots = function(year, codemodule, codeinstance, codeacti, callbackSuccess, callbackFailure) {
      var url;
      url = this.settings.url_activityMeetingSlots(year, codemodule, codeinstance, codeacti);
      this.log("Getting " + url);
      return this.request.get(url, (function(_this) {
        return function(error, response) {
          var json;
          if (!error && response.statusCode === 200) {
            json = _this.formatRequest(response);
            _this.log("Succesfully get'd " + url);
            if (callbackSuccess) {
              return callbackSuccess(json);
            }
          } else {
            _this.log("Failed to get " + url);
            if (callbackFailure) {
              return callbackFailure(error, response);
            }
          }
        };
      })(this));
    };

    return EpitechAPIConnector;

  })();

  exports.EpitechAPIConnector = EpitechAPIConnector;

}).call(this);
